src/libraries/System.Text.Json/docs/KeyConverter_spec.md

```diff
@@ -0,0 +1,185 @@
+|------------------------------------------ |------------------------- |---------:|---------:|---------:|---------:|---------:|---------:|-------:|-------:|------:|----------:|
+|   ReadDictionary<Dictionary<Guid, Int32>> | DeserializeFromUtf8Bytes | 34.83 us | 0.669 us | 0.593 us | 34.83 us | 33.84 us | 36.17 us | 5.8045 |      - |     - |  23.84 KB |
+|  ReadDictionary<Dictionary<Int32, Int32>> | DeserializeFromUtf8Bytes | 26.39 us | 0.448 us | 0.419 us | 26.33 us | 25.79 us | 27.43 us | 3.3389 |      - |     - |  13.82 KB |
+| ReadDictionary<Dictionary<String, Int32>> | DeserializeFromUtf8Bytes | 21.73 us | 0.378 us | 0.336 us | 21.77 us | 21.20 us | 22.29 us | 4.0248 | 0.1750 |     - |  16.77 KB |
+
+
+
+# Prior-art
+## Newtonsoft.Json
+ 
+### On write:
+
+* if the `TKey` is a concrete primitive type*:
+  * it calls `Convert.ToString()`
+  except for the next types:
+    * DateTime (uses `DateFormatHandling` specified in options)
+    * DateTimOffset
+    * Double (uses `double.ToString("R")`) // 'R' stands for round-trip
+    * Single
+    * Enum (uses an internal helper method)
+ 
+* If the `TKey` is `object` or non-primitive.
+  * it calls the `TypeConverter` of the `TKey` runtime type.
+  Except for :
+    * `Type`, which returns the `AssemblyQualifiedName`.
+  * If the type does not have a `TypeConverter`, it calls `ToString()` on the `TKey` instance.
+
+ 
+\* A *primitive type* is a value cataloged as such by Json.Net from [this list](https://github.com/JamesNK/Newtonsoft.Json/blob/a31156e90a14038872f54eb60ff0e9676ca4a0d8/Src/Newtonsoft.Json/Utilities/ConvertUtils.cs#L119-L168). 
+  
+### On read:
+
+* If the `TKey` is a concrete type.
+  * If is a primitive that implements `IConvertible`:
+    * Calls `Convert.ChangeType(propertyName, concreteType)` But first tries to manually convert on these types:
+      * Enum
+      * DateTime
+      * BigInteger
+    * If the type does not implement `IConvertible`:
+      * It tries to manually convert or cast to the concrete type using a few custom helper methods.
+* If the `TKey` is `object`, the entries' keys will be of type `string` if they are quoted (Newtonsoft supports unquoted properties).
+
+## Utf8Json
+Supported types:
+* String
+* Numerics (int, float, etc.)
+* Enum
+* Guid
+* Boolean
+* Nullables of all the above types
+
+## Jil
+Supported types:
+* String
+* Numerics (int, long, etc. Note: does not support floating point types)
+* Enum
+
+# Notes
+1. `DictionaryKeyPolicy` will apply to the resulting string of the non-string types.
+1. Should we provide a way to allow users to customize the `EnumKeyConverter` behavior, as it is done in `JsonStringEnumConverter`?
+As of now `KeyConverter`s are meant to be internal types, to enable the previously described behavior we either pass the options through `JsonSerializerOptions` or through an attribute.
```

Comment:
Do you imagine the work to make the type public should be scoped for 5.0? This initial support doesn't provide a way for users to customize (de)serialization of non-string dictionary keys. Is this acceptable for most users?

Reply:
As far as I have seen while writing the code, it does perfect sense to expose these types at some point, maybe you can make fit the effort needed for that into the 5.0 JSON roadmap.

Reply:
I think we should avoid scope creep and keep extension point + making it public out of the feature for now.

> Is this acceptable for most users?

That should be the goal by the default/built-in behavior. Let's say 80% folks are successful with it.